### ISSUE TITLE
Output variables fail to evaluate for Azure account fields

### ISSUE THREAD

**Issue by thebigmoosey** [OctopusDeploy/Issues#3724]
If you try to use an output variable as a custom expression on an Azure account field, it fails when you try to deploy the release:

For example, if you setup a script step called "set output var" that sets the output variable:
```
Set-OctopusVariable -name "AzureSubscriptionId" -value "[some valid azure account id here]"
```

Then reference this output variable as a custom expression on an Azure step, as follows:
```
#{Octopus.Action[set output var].Output.AzureSubscriptionId}
```

When you deploy the release, you receive this error message:
```
The expression #{Octopus.Action[set output var].Output.AzureSubscriptionId} could not be expanded to an Account ID. Make sure you have created the variables in this expression and scoped them correctly for the Test environment.
```

We also tried to "trick it" by creating a project-level variable that referenced the output variable instead, then using the project-variable in the Azure step, but the same error occurs.

We'll need to investigate why the account variable is not properly evaluating output variables, because there's a valid use case for switching between different Azure accounts in your deployments (based on scripting logic to set output variables).

Source: http://help.octopusdeploy.com/discussions/problems/56678
Related (possible duplicate): OctopusDeploy/Issues#3856
Related: https://github.com/OctopusDeploy/OctopusDeploy/issues/1711

**Comment by thebigmoosey** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Output variables fail to evaluate for Azure account fields](https://trello.com/c/Sc253mDj/256-output-variables-fail-to-evaluate-for-azure-account-fields)

**Comment by jrkd** [comment]
These trello links 404 (see also https://github.com/OctopusDeploy/Issues/issues/3856 for a related issue and additional trello 404)

**Comment by thebigmoosey** [comment]
@jrkd We use Trello internally (privately) as a workflow system for developers. The Trello links attach automatically when we've linked a GitHub issue to them (so we can keep everything linked together and find our way to the GitHub issue directly from Trello).

**Comment by ColinM9991** [comment]
I reported the issue linked above (#3856).

Is the piece of code that handles these principals closed source?

I'm happy to take a look and see if I can Pull Request a fix for this issue, as we have had to opt for "Blue/Green" approach in work to deploy to Azure in parallel, and that lifecycle doesn't accurately describe the cloud region.

**Comment by thebigmoosey** [comment]
@ColinM9991 I just took a loot, and it seems that this variable and account evaluation is not in one of our open-source repositories, so unfortunately you'll have to wait until the issue is resolved.

**Comment by ColinM9991** [comment]
@thebigmoosey Thanks for the update.

Is there visibility anywhere on where this is prioritized in the developers' backlog? As previously mentioned, this makes the "Cloud Regions" unusable for us and, to me, is quite a considerable bug.

**Comment by thebigmoosey** [comment]
@ColinM9991 We agree, and sorry we can't be of more help in terms of estimates, but this issue is in our pipeline and will be actioned by our cloud team.

We're currently reviewing the whole concept of "Cloud Regions". Our [specs repo for Resources](https://github.com/OctopusDeploy/Specs/blob/master/Resources/index.md) discusses the directions we're considering for a better cloud-first approach (this is an area that touches on accounts), so in terms of priorities, these variable issues will be better solved once we understand where cloud regions fit in this new model.

**Comment by octoreleasebot** [comment]
Release Note: Azure account verification is now deferred until the step runs, which means output variables from earlier steps can be used to resolve the account

**Comment by mchute84** [comment]
@octoreleasebot What version is this deferred in?

**Comment by matt-richardson** [comment]
@mchute84 - the fix was released on the 3rd of Jan. The "deferred" part just refers to the when we execute the verification (at execution time, rather than before the deployment begins)
